awscli/customizations/cloudformation/modules.py

```diff
@@ -0,0 +1,525 @@
+        """
+        if not isinstance(v, str):
+            msg = f"Ref should be a string: {v}"
+            raise exceptions.InvalidModuleError(msg=msg)
+
+        found = self.find_ref(v)
+        if found is not None:
+            d[n] = found
+
+    def find_ref(self, v):
+        """
+        Find a Ref.
+
+        A Ref might be to a module Parameter with a matching parent
+        template Property, or a Parameter Default. It could also
+        be a reference to another resource in this module.
+
+        :return The referenced element or None
+        """
+        # print(f"find_ref {v}, props: {self.props}")
+        if v in self.props:
+            if v not in self.params:
+                # The parent tried to set a property that doesn't exist
+                # in the Parameters section of this module
+                msg = f"{v} not found in module Parameters: {self.source}"
+                raise exceptions.InvalidModuleError(msg=msg)
+            return self.props[v]
+
+        if v in self.params:
+            param = self.params[v]
+            if DEFAULT in param:
+                # Use the default value of the Parameter
+                return param[DEFAULT]
+            msg = f"{v} does not have a Default and is not a Property"
+            raise exceptions.InvalidModuleError(msg=msg)
+
+        for k in self.resources:
+            if v == k:
+                # Simply rename local references to include the module name
+                return {REF: self.name + v}
+
+        return None
+
+    # pylint: disable=too-many-branches,unused-argument
+    def resolve_sub(self, v, d, n):
+        """
+        Parse the Sub string and break it into tokens.
+
+        If we can fully resolve it, we can replace it with a string.
+
+        Use the same logic as with resolve_ref.
+        """
+        # print(f"resolve_sub v: {v}, d: {d}, n: {n}")
+        words = parse_sub(v, True)
+        sub = ""
+        need_sub = False
+        for word in words:
+            if word.t == WordType.STR:
+                sub += word.w
+            elif word.t == WordType.AWS:
+                sub += "${AWS::" + word.w + "}"
+                need_sub = True
+            elif word.t == WordType.REF:
+                resolved = "${" + word.w + "}"
+                need_sub = True
+                found = self.find_ref(word.w)
+                if found is not None:
+                    if isinstance(found, str):
+                        need_sub = False
+                        resolved = found
+                    else:
+                        if REF in found:
+                            resolved = "${" + found[REF] + "}"
+                        elif SUB in found:
+                            resolved = found[SUB]
+                sub += resolved
+            elif word.t == WordType.GETATT:
+                need_sub = True
+                tokens = word.w.split()
+                if len(tokens) != 2:
```

Comment:
If my assumption is that this is a split on `.` for a GetAtt string in a Sub its worth noting that there can be extra dots here. I think your check would be better for < 2

Reply:
What's an example of this?

Reply:
[AWS::RDS::DBInstance](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-rds-dbinstance.html#aws-resource-rds-dbinstance-return-values)

```
CertificateDetails.CAIdentifier
    The CA identifier of the CA certificate used for the DB instance's server certificate.

CertificateDetails.ValidTill
    The expiration date of the DB instance’s server certificate.
```

[Nested stack Outputs](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-cloudformation-stack.html#aws-resource-cloudformation-stack-return-values) are another example

Reply:
Oh well then all of my `GetAtt` code is broken probably. I forgot about those. Why couldn't they just concatenate them?

Reply:
I think the rest of the code still works.  Like with the parser you can do `value = node.value.split(".", 1)` which will just split on the first occurrence.  The result is also a list of two items its just that the second item can have more `.`s